layers/mpls_echo.go

```diff
@@ -0,0 +1,550 @@
+	}
+
+	return &MPLSEchoTLV{
+		Type:  MPLSEchoTLVTypeTargetFECStack,
+		Value: byteRepresentation,
+	}, nil
+}
+
+// DecodeFromTLV decodes a given TLV and populates the TargetFECStackValue's fields with that data.
+func (s *TargetFECStackValue) DecodeFromTLV(tlv *MPLSEchoTLV) error {
+	if tlv.Type != MPLSEchoTLVTypeTargetFECStack {
+		return fmt.Errorf(
+			"TargetFECStackValue decoding error - type mismatch: expected %v, got %v",
+			MPLSEchoTLVTypeTargetFECStack,
+			tlv.Type)
+	}
+
+	s.FECs = nil
+	currentTLVStartOffset := 0
+	valueByteCount := len(tlv.Value)
+	for currentTLVStartOffset < valueByteCount {
+		currentTLV := &MPLSEchoTLV{}
+		numBytesDecoded, err := currentTLV.DecodeFromBytes(tlv.Value[currentTLVStartOffset:])
+		if err != nil {
+			return err
+		}
+
+		var newFEC MPLSEchoValue
+		switch currentTLV.Type {
+		case MPLSEchoSubTLVTypeTargetFECStackLDPPrefixIPv4:
+			fallthrough
+		case MPLSEchoSubTLVTypeTargetFECStackLDPPrefixIPv6:
+			newFEC = &LDPPrefixFECValue{}
+			err = newFEC.DecodeFromTLV(currentTLV)
+			if err != nil {
+				return err
+			}
+
+		default:
+			return fmt.Errorf("TargetFECStackValue decoding error - unknown sub-type: %v", currentTLV.Type)
+		}
+
+		s.FECs = append(s.FECs, newFEC)
+		currentTLVStartOffset += numBytesDecoded
+	}
+
+	if currentTLVStartOffset != valueByteCount {
+		return fmt.Errorf(
+			"TargetFECStackValue decoding error - sub-TLVs don't match the valid data bounds (Total Valid Bytes = %v, Last TLV End Offset = %v)",
+			valueByteCount,
+			currentTLVStartOffset)
+	}
+
+	return nil
+}
+
+// LayerType returns gopacket.LayerTypeMPLSEcho.
+func (m *MPLSEcho) LayerType() gopacket.LayerType {
+	return LayerTypeMPLSEcho
+}
+
+// DecodeFromBytes decodes the given bytes into this layer.
+func (m *MPLSEcho) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+	if len(data) < 32 {
+		df.SetTruncated()
+		return fmt.Errorf("Invalid MPLSEcho content - length %v less than 32 bytes", len(data))
+	}
+
+	totalTLVBytes := len(data) - 32
+	if totalTLVBytes%4 != 0 {
+		return fmt.Errorf("Invalid MPLSEcho content - TLVs aren't 4-octet aligned (TLV bytes = %v)", totalTLVBytes)
+	}
+
+	m.VersionNumber = MPLSEchoVersion(binary.BigEndian.Uint16(data[0:2]))
+	m.GlobalFlags = MPLSEchoGlobalFlags(binary.BigEndian.Uint16(data[2:4]))
+	m.MessageType = MPLSEchoMessageType(data[4])
+	m.ReplyMode = MPLSEchoReplyMode(data[5])
+	m.ReturnCode = MPLSReplyReturnCode(data[6])
+	m.ReturnSubcode = data[7]
+	m.SenderHandle = binary.BigEndian.Uint32(data[8:12])
+	m.SequenceNumber = binary.BigEndian.Uint32(data[12:16])
+	m.TimestampSentSeconds = binary.BigEndian.Uint32(data[16:20])
+	m.TimestampSentSecondsFraction = binary.BigEndian.Uint32(data[20:24])
+	m.TimestampReceivedSeconds = binary.BigEndian.Uint32(data[24:28])
+	m.TimestampReceivedSecondsFraction = binary.BigEndian.Uint32(data[28:32])
+
+	currentTLVStartOffset := 32
+	for currentTLVStartOffset < len(data) {
+		currentTLV := &MPLSEchoTLV{}
```

Comment:
See comment in the data structure (use preallocation like options in IP4 or TCP) and I think not using a pointer would be better here.
Memory allocations should be avoided.